### PR TITLE
Separate dts generator for ace-code and ace-builds

### PR DESCRIPTION
**Pull request by mkslanc**
*Issue #, if available:* #4640

*Description of changes:*
This pull request introduces changes that enable the generation of a comprehensive TypeScript declaration file, designed to be utilized as a public API. These modifications ensure that the generated declaration file accurately reflects the structure of both `ace-builds` and `ace-code` packages, providing a seamless integration for developers.

**Key Features:**

- Unified Declaration: Consolidates type definitions across ace-builds and ace-code, facilitating easier maintenance and usage.
- Project Structure Compliance: Adheres to the respective project structures, ensuring that the generated file is both accurate and functional.
- All internal statements (started with $ or marked as `private` or `@internal` in jsDoc comment) would be filtered from final declaration

**Disclaimer:** For ace-builds, all internal module paths are modified to follow the `ace-builds-internal/module` structure, which could be used as completion items in IDE's. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by akoreman** [comment]
Cool stuff! 

It seems that by automatically generating the declaration files, we're potentially adding stuff to our public contract (the `*.d.ts` files) that were previously unexposed to users. Did we go through the new declaration files and check that everything newly exposed is properly marked as `private`?

**Comment by mkslanc** [comment]
> Cool stuff!
> 
> It seems that by automatically generating the declaration files, we're potentially adding stuff to our public contract (the `*.d.ts` files) that were previously unexposed to users. Did we go through the new declaration files and check that everything newly exposed is properly marked as `private`?

I am removing all nodes marked as `@private` (`/**@private*/`) and all identifiers starting with $ (as this format is used to mark private methods/properties). Therefore, if something needs to be removed, we can easily add the @private modifier in the documentation comment.

**Comment by mkslanc** [comment]
I have generated a list of newly exposed entities, which you can find [here](https://github.com/user-attachments/files/15776609/diff-between-old-new.txt). This should help in deciding which ones to make private.

**Disclaimer**: This comparison only includes entities within the Ace namespace. If whole class was missing in old `ace.d.ts` it would have only top-level identifier.

**Comment by InspiredGuy** [comment]
@mkslanc there were some additions to the declaration files in the PRs merged lately, could you please rebase/merge and resolve conflicts?

**Comment by mkslanc** [comment]
> @mkslanc there were some additions to the declaration files in the PRs merged lately, could you please rebase/merge and resolve conflicts?

Done! + I improved filtering mechanism, now any node (function/property) marked with `/** @internal */` would be removed from public API

File: ace.d.ts [hunk 1]
```diff
@@ -1,1170 +1,1003 @@
...
+        }): import("ace-code").Ace.AppConfig;
+        resetOptions(obj: any): void;
+        setDefaultValue(path: string, name: string, value: any): boolean;
+        setDefaultValues(path: string, optionHash: {
+            [key: string]: any;
+        }): void;
+        setMessages(value: any, options?: {
+            placeholders?: "dollarSigns" | "curlyBrackets";
+        }): void;
+        nls(key: string, defaultString: string, params?: {
```
Comment (nightwing) [inline ace.d.ts:958]: There is a conflict between this and the definition on line 72? why do we even have this duplication and the version returning any?

Comment (mkslanc) [reply ace.d.ts:958]: Corrected differences between old config declaration in Ace namespace and generated one. (By using type alias to actual config). Copying same config type into `const config` happens on initial declaration generation, the only solution I could suggest is to manually find config node in ace.js on generator stage and replacing it with `Ace.Config` one

**Pull request merged**